Advent 2018: How to Send and Receive SMS: Implementing a GSM Protocol in Go (#292)
gopheracademy/gopheracademy-web: content/advent-2018/how-to-send-and-receive-sms.md

### alert operation

Allows us to send pings to the SMSC.

### submit short message operation

Allows us to send mobile terminating SMS.

### delivery notification operation

Sent by the SMSC to the client as a delivery status receipt indicating if the SMS message was successfully sent or not.

### delivery short message operation

Sent by the SMSC to the client on behalf of a subscriber's mobile originating SMS message.

## Implementation

We can treat UCP like a traditional client-server protocol. After etablishing a TCP connection, we send UCP requests with a sequence number(called transaction reference number in the spec) from 00 to 99 and the SMSC will respond with a UCP response message synchronously. However the SMSC can also send UCP requests, just like in the case of `delivery notification operation` and `delivery short message operation`. We also need to periodically send keepalive pings to the SMSC so that it wont treat the connection as stale and disconnect us. You can read the protocol specification [here](http://documents.swisscom.com/product/1000174-Internet/Documents/Landingpage-Mobile-Mehrwertdienste/UCP_R4.7.pdf).

> Lets start with a `Client` struct containing the login credentials to the SMSC.

Reviewer: In the "Conclusion" section you're saying 

> If you work on the telco field and have an access to an SMSC, feel free to try [..]

I feel it will be nice to add more context about the project we're going to build. As an example, the `Client` below has address and credentials. But how one can get them? I think a paragraph that describes the requirements for the `Client` will make the post more compleat.

Author: I'll add a section `Demo` wherein I'll use an SMSC simulator. I'll explore the UCP packets in Wireshark too. I'll finish up the simulator first, then add the github repo link so that the readers can follow along even though they wont send any real SMS.

          writer.Write(deliverySmAckPacket([]byte(refNum), sysmsg))
          writer.Flush()
          
          var incomingMsg deliverMsgPart
          incomingMsg.sender = sender
          incomingMsg.receiver = recvr
          incomingMsg.message = msg
          incomingMsg.msgID = msgID
          // further processing    
        }
      }
    }
  }()
}
```

`deliverMsgPart` is a struct that contains the neccessary parts to concatenate and decode the partial incoming mobile originating message.

```go
// deliverMsgPart represents a deliver sm message part
type deliverMsgPart struct {
  currentPart int
  totalParts  int
  refNum      int
  sender      string
  receiver    string
  message     string
  msgID       string
  dcs         string
}
```

To handle multi-part mobile originating SMS, we send partial mobile originating messages to `deliverMsgPartCh` channel and complete mobile originating messages to `deliverMsgCompleteCh` channel.

```go

// Reads all deliver sm messages(mobile-originating messages) 
// from the deliverMsgCh channel.
func readDeliveryMsg(/*....*/) {
  wg.Add(1)
  go func() {
    defer wg.Done()

>       for {

Reviewer: I think the indentation here is wrong, doesn't it? Maybe you could double-check if all code snippets in the post were `gofmt`-ed.

            // handle multi-part mobile originating message
            // get the total message parts in the xser data
            msgPartsLen := xserUdh[len(xserUdh)-4 : len(xserUdh)-2]
            // get the current message part in the xser data
            msgPart := xserUdh[len(xserUdh)-2:]
            // get message part reference number
            msgRefNum := xserUdh[len(xserUdh)-6 : len(xserUdh)-4]
            // convert hexstring to integer
            msgRefNumInt, _ := strconv.ParseInt(msgRefNum, 16, 0)
            msgPartsLenInt, _ := strconv.ParseInt(msgPartsLen, 16, 64)
            msgPartInt, _ := strconv.ParseInt(msgPart, 16, 64)
            incomingMsg.currentPart = int(msgPartInt)
            incomingMsg.totalParts = int(msgPartsLenInt)
            incomingMsg.refNum = int(msgRefNumInt)
            // send to partial channel
            deliverMsgPartCh <- incomingMsg 
          } else {
          	// handle mobile originating message with only 1 part
          	// send the incoming message to the complete channel
          	deliverMsgCompleteCh <- incomingMsg
          }
        }
      }
    }
  }()
}
```

> The spawned a goroutine `readPartialDeliveryMsg()` will receive from`deliverMsgPartCh` channel and concatenate the incoming mobile originating message parts while the other spawned goroutine `readCompleteDeliveryMsg()` will receive from `deliverMsgCompleteCh` channel and execute the callback for mobile originating messages.

Reviewer: This sounds a bit odd to me. How about the following:

```suggestion
The goroutine spawned in `readPartialDeliveryMsg` will read from `deliverMsgPartCh` channel and concatenate the incoming mobile originating message parts. The goroutine spawned in `readCompleteDeliveryMsg` will receive from `deliverMsgCompleteCh` channel and execute the callback for mobile originating messages.
```

  for i := 0; i < len(msgParts); i++ {
    sendPacket := encodeMessage(c.nextRefNum(), sender, receiver, msgParts[i], msgType,
      c.GetBillingID(), refNum, i+1, len(msgParts))
    c.writer.Write(sendPacket)
    c.writer.Flush()
    select {
    case fields := <-c.submitSmRespCh:
      ack := fields[ackIndex]
      if ack == negativeAck {
        errMsg := fields[len(fields)-errMsgOffset]
        errCode := fields[len(fields)-errCodeOffset]
        return ids, &UcpError{errCode, errMsg}
      }
      id := fields[submitSmIdIndex]
      ids[i] = id
    case <-time.After(c.timeout):
      return ids, &UcpError{errCodeTimeout, "Network time-out"}
    }
  }
  return ids, nil
}
```
`getMessageType` determines whether the message contains plain GSM 7-bit characters or Unicode characters.

`getMessageParts` splits the message into multiple parts in case it's a multi-part message.

`encodeMessage` takes care of creating a valid `submit short message orperation` request packet with the appropriate reference number. It handles text encoding to [UCS2](https://en.wikipedia.org/wiki/Universal_Coded_Character_Set) for unicode messages as well as masking the sender name. 

> To get the response from the SMSC, we use a `select` statement between the `submitSmRespCh` channel and a given timeout.

Reviewer: ```suggestion
To get the response from the SMSC, we use `select` statement, that blocks until the data from the `submitSmRespCh` channel can be read or a given timeout occurred.
```
